File: app/src/main/java/com/bestxty/dl/demo/adapter/TaskAdapter.java
package com.bestxty.dl.demo.adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.Button;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.bestxty.dl.demo.R;
import com.bestxty.dl.demo.bean.Task;

import java.util.List;

/**
 * @author swjjx
 *         Created by swjjx on 2016/12/20.
 */

public class TaskAdapter extends BaseAdapter {

    private List<Task> taskList;
    private LayoutInflater layoutInflater;

    public TaskAdapter(Context context, List<Task> taskList) {
        this.layoutInflater = LayoutInflater.from(context);
        this.taskList = taskList;
    }

    @Override
    public int getCount() {
        return taskList.size();
    }

    @Override
    public Object getItem(int i) {
        return taskList.get(i);
    }

    @Override
    public long getItemId(int i) {
        return i;
    }

    @Override
    public View getView(int i, View view, ViewGroup viewGroup) {
        TaskViewHolder holder = null;
        if (view == null) {
            view = layoutInflater.inflate(R.layout.item_task, viewGroup, false);
            holder = new TaskViewHolder();
            holder.title = (TextView) view.findViewById(R.id.text_title);
            holder.desc = (TextView) view.findViewById(R.id.text_desc);
            holder.button = (Button) view.findViewById(R.id.btn_task);
            holder.progressBar = (ProgressBar) view.findViewById(R.id.pb_progress);
            view.setTag(holder);
        } else {
            holder = (TaskViewHolder) view.getTag();
        }

        Task task = taskList.get(i);
        holder.title.setText(task.getTitle());
        holder.desc.setText(task.getDesc());
        holder.button.setText(getTaskState(task.getState()));
        if (task.getState() == Task.STATE_READY) {
            holder.progressBar.setVisibility(View.GONE);
        } else {
            holder.progressBar.setVisibility(View.VISIBLE);
        }

        return view;
    }

    private String getTaskState(int state) {
        switch (state) {
            case Task.STATE_READY:
                return "Start";
            case Task.STATE_DOWNING:
                return "DOWN";
            case Task.STATE_PAUSE:
                return "PAUSE";
            case Task.STATE_DONE:
                return "DONE";
        }
        return "ERROR";
    }

    public static class TaskViewHolder {
        TextView title;
        TextView desc;
       public Button button;
       public ProgressBar progressBar;
    }
}

File: app/src/main/java/com/bestxty/sault/demo/adapter/TaskAdapter.kt
package com.bestxty.sault.demo.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.Button
import android.widget.ProgressBar
import android.widget.TextView
import com.bestxty.sault.demo.R
import com.bestxty.sault.demo.bean.Task

/**
 * @author swjjx
 *         Created by swjjx on 2016/12/20.
 */

class TaskAdapter(context: Context, private var taskList: List<Task>) : BaseAdapter() {

    private var layoutInflater: LayoutInflater = LayoutInflater.from(context)

    override fun getCount(): Int = taskList.size

    override fun getItem(position: Int): Any = taskList[position]

    override fun getItemId(position: Int): Long = position.toLong()


    override fun getView(position: Int, convertView: View?, parent: ViewGroup?): View {
        val holder: TaskViewHolder
        val view: View
        if (convertView == null) {
            view = layoutInflater.inflate(R.layout.item_task, parent, false)
            holder = TaskViewHolder()
            holder.title = view.findViewById(R.id.text_title) as TextView
            holder.desc = view.findViewById(R.id.text_desc) as TextView
            holder.button = view.findViewById(R.id.btn_task) as Button
            holder.progressBar = view.findViewById(R.id.pb_progress) as ProgressBar
            view.tag = holder
        } else {
            view = convertView
            holder = view.tag as TaskViewHolder
        }
        val task: Task = taskList[position]
        holder.title.text = task.title
        holder.desc.text = task.desc
        holder.button.text = getTaskState(task.state)
        if (task.state == Task.STATE_READY
                || task.state == Task.STATE_DONE) {
            holder.progressBar.visibility = View.GONE
        } else {
            holder.progressBar.visibility = View.VISIBLE
        }
        return view
    }


    private fun getTaskState(state: Int): String {
        when (state) {
            Task.STATE_READY ->
                return "Start"
            Task.STATE_DOWNING ->
                return "DOWN"
            Task.STATE_PAUSE ->
                return "PAUSE"
            Task.STATE_DONE ->
                return "DONE"
        }
        return "ERROR"
    }

    class TaskViewHolder {
        lateinit var title: TextView
        lateinit var desc: TextView
        lateinit var button: Button
        lateinit var progressBar: ProgressBar
    }

}
